test/pylib/scylla_cluster.py

```diff
@@ -0,0 +1,792 @@
+        if self.start_exception:
+            raise self.start_exception
+
+        for server in self.running.values():
+            server.write_log_marker(f"------ Starting test {name} ------\n")
+            server.take_log_savepoint()
+
+    def after_test(self, name) -> None:
+        """Check that the cluster is still alive and the test
+        hasn't left any garbage."""
+        assert(self.start_exception is None)
+        if self._get_keyspace_count() != self.keyspace_count:
+            raise RuntimeError("Test post-condition failed, "
+                               "the test must drop all keyspaces it creates.")
+        for server in itertools.chain(self.running.values(), self.stopped.values()):
+            server.write_log_marker(f"------ Ending test {name} ------\n")
+
+    def update_last_seed(self, removed_host: str) -> None:
+        """Update last seed when removing a host"""
+        if self.last_seed == removed_host:
+            self.last_seed = list(self.running.keys())[0] if self.running else None
+
+    async def server_stop(self, server_id: str, gracefully: bool) -> Tuple[bool, str]:
+        """Stop a server. No-op if already stopped."""
+        logging.info("Cluster %s stopping server %s", self, server_id)
+        if server_id in self.stopped:
+            return True, f"Server {server_id} already stopped"
+        if server_id in self.removed:
+            return False, f"Server {server_id} removed"
+        if server_id not in self.running:
+            return False, f"Server {server_id} unknown"
+        self.is_dirty = True
+        server = self.running.pop(server_id)
+        if gracefully:
+            await server.stop_gracefully()
+        else:
+            await server.stop()
+        self.update_last_seed(server.host)
+        self.stopped[server_id] = server
+        return True, f"Server {server_id} stopped"
+
+    async def server_start(self, server_id: str) -> Tuple[bool, str]:
+        """Start a stopped server"""
+        logging.info("Cluster %s starting server", self)
+        if server_id in self.running:
+            return True, f"Server {server_id} already started"
+        if server_id in self.removed:
+            return False, f"Server {server_id} removed"
+        if server_id not in self.stopped:
+            return False, f"Server {server_id} unknown"
+        self.is_dirty = True
+        server = self.stopped.pop(server_id)
+        if self.last_seed is None:
+            self.last_seed = server.host
+        server.seed = self.last_seed
+        await server.start()
+        self.running[server_id] = server
+        return True, f"Server {server_id} started"
+
+    async def server_restart(self, server_id: str) -> Tuple[bool, str]:
+        """Restart a running server"""
+        logging.info("Cluster %s restarting server %s", self, server_id)
+        if server_id in self.running:
+            self.is_dirty = True
+            server = self.running.pop(server_id)
```

Comment:
here and elsewhere in this class it bothers me that we may leak the server if the line below throws (stop_gracefully(), start()). This is a rest api call, and it is legitimate to fail, the failure may be well ignored by the test, or the test may be testing for failure. I think we should not leak servers in any case, from any of this list. Do we really need these lists, maybe the server internal 'state' variable (running, stopped, removed) would be enough?

Reply:
API commands need to get a server by id so having a plain list and state inside the Server object is not good enough.

It's true this might leak. I'll fix later.